src/de.rs

```diff
@@ -1362,70 +640,63 @@ where
 /// starts to parse the bytes, before we could interfere. If the data only includes a CID, we are
 /// parsing over the tag to determine whether it is a CID or not and go from there.
+//struct CidDeserializer<'a, 'b, R: dec::Read<'b>>(&'a mut Deserializer<'b, R>);
 struct CidDeserializer<'a, R>(&'a mut Deserializer<R>);
 
-//impl<'de, 'a: 'de, R> de::Deserializer<'de> for &'a mut CidDeserializer<'a, R>
-impl<'de, 'a, R> de::Deserializer<'de> for &'a mut CidDeserializer<'a, R>
-where
-    R: Read<'de>,
-{
-    type Error = Error;
+impl<'de, 'a, R: dec::Read<'de>> de::Deserializer<'de> for &'a mut CidDeserializer<'a, R> {
+    type Error = DecodeError<R::Error>;
 
-    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value> {
-        Err(Error::message("Only bytes can be deserialized into a CID"))
+    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value, Self::Error> {
+        Err(de::Error::custom(
+            "Only bytes can be deserialized into a CID",
+        ))
     }
 
     #[inline]
-    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
-        // Match on the major type, it must be a byte string (major type 2)
-        let len = match self.0.parse_u8()? {
-            // CIDs always have a `0x00` prefix, hence they cannot be zero sized.
-            0x40 => return Err(self.0.error(ErrorCode::LengthOutOfRange)),
-            byte @ 0x41..=0x57 => usize::try_from(byte - 0x40)
-                .map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?,
-            0x58 => {
-                let len = self.0.parse_u8()?;
-                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
-            }
-            0x59 => {
-                let len = self.0.parse_u16()?;
-                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
-            }
-            0x5a => {
-                let len = self.0.parse_u32()?;
-                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
-            }
-            0x5b => {
-                let len = self.0.parse_u64()?;
-                usize::try_from(len).map_err(|_| self.0.error(ErrorCode::LengthOutOfRange))?
-            }
-            _ => return Err(self.0.error(ErrorCode::UnexpectedCode)),
-        };
-
-        match self.0.read.read(len)? {
-            EitherLifetime::Long(buf) | EitherLifetime::Short(buf) => {
-                // In DAG-CBOR the CID is prefixed with a null byte, strip that off.
-                visitor.visit_bytes(&buf[1..])
+    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        let byte = peek_one(&mut self.0.reader)?;
+        match dec::if_major(byte) {
+            major::BYTES => {
+                // CBOR encoded CIDs have a zero byte prefix we have to remove.
```

Comment:
why?

Reply:
Why CBOR encodes CIDs with a zero byte prefix? This has historic reasons. CIDs are specified as being Multibase encoded. And the prefix `0x00` stands for "it's not encoded, use it as is". Though these days I don't see anyone considering CIDs being Multibase encoded, but rather speaks of "Multibase encoded CIDs" in case they are. This is the reason, why the CID implementation we are using (and any other CID implementation I know of) doesn't contain the `0x00` when returning a CID.

Reply:
😭